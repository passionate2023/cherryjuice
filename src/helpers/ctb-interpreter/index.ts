import { translateAttributesToHtmlAndCss } from './steps/translate-attributes-to-html-and-css';
import { parseString } from 'xml2js';
import { flattenIntoLines } from './steps/flatten-into-lines';
import { groupNodesByLine } from './steps/group-nodes-by-line';
import { insertOtherTables } from './steps/insert-other-tables';

const parse = xml =>
  new Promise((resolve, reject) => {
    parseString(xml, async function(err, result) {
      err ? reject(err) : resolve(result);
    });
  });

const parseRichText = async ({
  nodeTableXml,
  otherTables = {},
  meta: { name },
  options: { stringify }
}) => {
  const parsedXml = await parse(nodeTableXml);
  // @ts-ignore
  const richText = parsedXml.node.rich_text;
  console.log('pre insert-other-tables', richText);
  const xml = insertOtherTables({ xml: richText, otherTables });
  console.log('pre css-html translator', xml);
  const translated = xml.map(node => {
    if (typeof node === 'object') {
      node.$$ = translateAttributesToHtmlAndCss(node.$);
    }
    return node;
  });
  console.log('pre separation', translated);
  const separated = flattenIntoLines(translated);
  console.log('pre splitting', separated);
  const split = groupNodesByLine(separated);
  console.log('pre resolving', split);
  return stringify ? JSON.stringify(split) : split;
};
export { parseRichText };
