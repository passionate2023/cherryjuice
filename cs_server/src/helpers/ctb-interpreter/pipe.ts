import { translateAttributesToHtmlAndCss } from './steps/translate-attributes-to-html-and-css';
import { flattenIntoLines } from './steps/flatten-into-lines';
import { groupNodesByLine } from './steps/group-nodes-by-line';
import { insertOtherTables } from './steps/insert-other-tables';
import { parseXml } from './helpers';
import { compose } from 'ramda';
import { fixCharacters } from './steps/fix-characters';
import { escapeHtml } from './helpers/escape-html';

const tap = label => val => (console.log(label, val), val);
const stringifierPipe = compose(
  fixCharacters.replaceTabCharacter,
  fixCharacters.replaceSpaceCharacter,
  JSON.stringify,
);
const processingPipe = otherTables =>
  compose(
    tap('two'),
    translateAttributesToHtmlAndCss,
    groupNodesByLine,
    flattenIntoLines,
    insertOtherTables(otherTables),
    fixCharacters.restoreOrphanWhiteSpace,
  );
const parseRichText = async ({
  nodeTableXml,
  otherTables = {},
  stringify = true,
}) => {
  nodeTableXml = fixCharacters.flagOrphanWhiteSpace(nodeTableXml);
  let richText = await parseXml({ xml: nodeTableXml }).then(
    ({ node: { rich_text } }) => rich_text,
  );
  const res = processingPipe(otherTables)(richText);
  return stringify ? stringifierPipe(res) : res;
};

export { parseRichText, processingPipe };
