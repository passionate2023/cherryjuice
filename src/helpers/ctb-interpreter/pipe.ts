import { translateAttributesToHtmlAndCss } from './steps/translate-attributes-to-html-and-css';
import { flattenIntoLines } from './steps/flatten-into-lines';
import { groupNodesByLine } from './steps/group-nodes-by-line';
import { insertOtherTables } from './steps/insert-other-tables';
import { parseXml } from './helpers';
import { compose } from 'ramda';
import { fixCharacters } from './steps/fix-characters';

const tap = label => val => (console.log(label, val), val);

const processingPipe = otherTables =>
  compose(
    fixCharacters.replaceTabCharacter,
    fixCharacters.replaceSpaceCharacter,
    JSON.stringify,
    tap('two'),
    groupNodesByLine,
    tap('one'),
    // @ts-ignore
    flattenIntoLines,
    translateAttributesToHtmlAndCss,
    insertOtherTables(otherTables),
    fixCharacters.restoreOrphanWhiteSpace
  );
const parseRichText = async ({ nodeTableXml, otherTables = {} }) => {
  nodeTableXml = fixCharacters.flagOrphanWhiteSpace(nodeTableXml);
  let richText = await parseXml({ xml: nodeTableXml }).then(
    ({ node: { rich_text } }) => rich_text
  );
  return processingPipe(otherTables)(richText);
};

export { parseRichText, processingPipe };
