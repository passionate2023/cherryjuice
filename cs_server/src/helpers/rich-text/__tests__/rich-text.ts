import { RichText } from '../index';
import { parseRichText } from '../../ctb-interpreter/pipe';
import { renderingIssues } from '../../ctb-interpreter/__tests__/__data__';
const renderingTestTemplate = async ({ txt, otherTables }) => {
  const res = RichText({
    richText: await parseRichText({
      nodeTableXml: txt,
      otherTables,
      stringify: false,
    }),
  });
  expect(res).toMatchSnapshot();
};

describe('rich-text snapshot tests', () => {
  test(renderingIssues.issue_01.name, () => {
    renderingTestTemplate(renderingIssues.issue_01);
  });
  test(renderingIssues.issue_02.name, () => {
    renderingTestTemplate(renderingIssues.issue_02);
  });
  test(renderingIssues.issue_03.name, () => {
    renderingTestTemplate(renderingIssues.issue_03);
  });
  test(renderingIssues.issue_04.name, () => {
    renderingTestTemplate(renderingIssues.issue_04);
  });
  test(renderingIssues.issue_05.name, () => {
    renderingTestTemplate(renderingIssues.issue_05);
  });
  test(renderingIssues.issue_06.name, () => {
    renderingTestTemplate(renderingIssues.issue_06);
  });
  test(renderingIssues.issue_07.name, () => {
    renderingTestTemplate(renderingIssues.issue_07);
  });
  test(renderingIssues.issue_08.name, () => {
    renderingTestTemplate(renderingIssues.issue_08);
  });
});
