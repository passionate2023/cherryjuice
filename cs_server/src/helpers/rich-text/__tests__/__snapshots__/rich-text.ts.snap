// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rich-text snapshot tests adjacent images overlap 1`] = `
"<article>- we can think of callbacks as a continuation, a secondary half that runs after a time-gap that separates it from the first half<br />&quot;w<div class=\\"rich-text__text\\" style=\\"color:#e6e6e6;\\">hen we refer to functions as continuations, what we mean is, we did the first half of our program now, and we wrap the second half of our program up in a callback</div>&quot;<br /><br />- the instructor suggests that callback hell doesnt actually have to do with indentation, as we could rewrite a &quot;callback hell&quot; without nesting<br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />the instructor suggests that t<div class=\\"rich-text__text\\" style=\\"color:#e6e6e6;\\">here are two conceptually deficient</div> problems with callbacks<br />
    </article>"
`;

exports[`rich-text snapshot tests empty newline ignored 1`] = `
"<article>the instructor suggests<br /><br />generators will solve<br /><br /><br />if a yield<br /><br />instead <br /><br /><br /><br />yield &gt; pause &gt; then &gt; it.next<br />
    </article>"
`;

exports[`rich-text snapshot tests monospace style not fully applied 1`] = `
"<article><small class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">i </code></small><small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">a </code></small><small class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">f </code></small><small class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">&#39;&#39;</code></small><small class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small><br /><small class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">import * </code></small><small class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">f </code></small><small class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">&#39;&#39;</code></small><small class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small><br />
    </article>"
`;

exports[`rich-text snapshot tests offset images 1`] = `
"<article><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">let </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">it </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">= </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">gen</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">()</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#c792ea;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">console</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#82aaff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">log</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#f78c6c;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">it</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#82aaff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">next</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#f78c6c;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">())</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#9a1197;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#c792ea;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">console</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#82aaff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">log</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#f78c6c;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">it</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#82aaff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">next</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#f78c6c;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(5))</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><br /><br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><br /><br /><br />example2<br /><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />(run is equivalent to it.next)<br />
    </article>"
`;

exports[`rich-text snapshot tests offset images 2`] = `
"<article><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\"> </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#ffffff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\"> 5</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">  </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#c792ea;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">let </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">y </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">= </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#f78c6c;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">0 </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#498b49;color:#89ddff;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">+ </code></small></em><br /><br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><br /><br /><br />example2<br /><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />(run is equivalent to it.next)<br />
    </article>"
`;

exports[`rich-text snapshot tests offset images, tables, and codebox 1`] = `
"<article>- fr<div class=\\"rich-text__text\\" style=\\"color:#ffffff;\\"> and </div><code class=\\"rich-text__text\\" style=\\"color:#ffffff;background-color:#2B2B2B;\\">split</code><div class=\\"rich-text__text\\" style=\\"color:#ffffff;\\">.</div><br />
    <table class=\\"rich-text__table\\" >
      <thead>
        <th >name</th><th >age</th>
      </thead>
      <tbody>
        <tr >
            <td >yacine</td><td >26</td>
          </tr><tr >
            <td >kamel</td><td >27</td>
          </tr><tr >
            <td >hachemi</td><td >26</td>
          </tr><tr >
            <td ></td><td ></td>
          </tr><tr >
            <td ></td><td ></td>
          </tr><tr >
            <td >amine</td><td >26</td>
          </tr>
      </tbody>
    </table>
  <br /><br /><br />&gt;regex functions<br />a) regex.test returns a boolean if the pattern matches<br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">let </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">str </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">= </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">\`Is this This?\`</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">console</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">log</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">/isd/g</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">test</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">str</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">))</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">; </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#546e7a;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">// false</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">console</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">log</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">/is/g</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">test</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">str</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">))</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">; </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#546e7a;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">// true</code></small></em><br /><br />b) regex.exec returns a regexArray<br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">let </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">str </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">= </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">\`Is this This?\`</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">let </code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">regex </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">= </code></small></em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">/is/g</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">console</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">log</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">regex</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">.</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#82aaff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">exec</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">(</code></small></em><em class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">str</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#f78c6c;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">))</code></small></em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">;</code></small></em><br /><em class=\\"rich-text__text\\" style=\\"color:#546e7a;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">// [ &#39;is&#39;, index: 5, input: &#39;Is this This?&#39;, groups: undefined ]</code></small></em><br /><br />unlike <small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">test</code></small>, the <small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">exec</code></small> method is stateful: <br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><br /><br />&gt;string functions<br />a) <small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">str.match</code></small><br /><em class=\\"rich-text__text\\" style=\\"color:#c792ea;\\">let </em><em class=\\"rich-text__text\\" style=\\";\\">str </em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;\\">= </em><em class=\\"rich-text__text\\" style=\\"color:#c3e88d;\\">\`Is this This?\`</em><em class=\\"rich-text__text\\" style=\\"color:#89ddff;\\">;</em><br /><br /><br />&gt;basic regex playground using <small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">str.replace</code></small><br /><code
      class=\\"rich-text__code\\"
      style=\\"max-width: 100px;
        min-height: 100px;
        width: 100px;
        display: inline-block\\" 
    >
      codebox0
    </code><br /><code
      class=\\"rich-text__code\\"
      style=\\"max-width: 200px;
        min-height: 200px;
        width: 200px;
        display: inline-block\\" 
    >
      codebox1
    </code><br /><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />*<br />&gt;my react alternative<br /><div style=\\" width:300px; height:300px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:300px; height:300px ;\\"
      />
    </div><br />- full code<br /><em class=\\"rich-text__text\\" style=\\"color:#89ddff;background-color:#2B2B2B;\\"><small class=\\"rich-text__text\\" style=\\";\\"><code class=\\"rich-text__text\\" style=\\";\\">i</code></small></em><br />
    </article>"
`;

exports[`rich-text snapshot tests tailing image not rendered 1`] = `
"<article>promises can be chained together<br />&gt;example<br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><br />the <small class=\\"rich-text__text\\" style=\\"background-color:#2B2B2B;\\"><code class=\\"rich-text__text\\" style=\\";\\">.then</code></small> on line 5 will &quot;magically&quot; wait for the promise from line 3<br /><br />&gt;example2<br /><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />
    </article>"
`;

exports[`rich-text snapshot tests tailing image not rendered 2`] = `
"<article>haha<br /><div style=\\" width:100px; height:100px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:100px; height:100px ;\\"
      />
    </div><div style=\\" width:200px; height:200px ; display: inline-block\\">
      <img src=\\"\\"
        alt=\\"\\"
        style=\\" width:200px; height:200px ;\\"
      />
    </div><br />
    </article>"
`;
